src/spec_tools/visitor.cljc

```diff
@@ -22,137 +23,153 @@
 (defn extract-form [spec]
   (if (seq? spec) spec (s/form spec)))
 
+(defn namespaced-name [key]
+  (if key
+    (if-let [nn (namespace key)]
+      (str nn "/" (name key))
+      (name key))))
+
+(defn unwrap
+  "Unwrap [x] to x. Asserts that coll has exactly one element."
+  [coll]
+  {:pre [(= 1 (count coll))]}
+  (first coll))
+
 (defn- spec-dispatch
-  [spec accept]
+  [spec accept options]
   (cond
     (or (s/spec? spec) (s/regex? spec) (keyword? spec))
     (let [form (s/form spec)]
       (if (not= form ::s/unknown)
         (if (seq? form)
           (normalize-symbol (first form))
-          (spec-dispatch form accept))
+          (spec-dispatch form accept options))
         spec))
     (set? spec) ::set
     (seq? spec) (normalize-symbol (first (strip-fn-if-needed spec)))
-    :else (normalize-symbol spec)))
+    (symbol? spec) (normalize-symbol spec)
+    :else (normalize-symbol (form/resolve-form spec))))
 
-(defn- expand-spec-ns [x]
-  (if-not (namespace x) (symbol "clojure.core" (name x)) x))
+(defmulti visit-spec spec-dispatch :default ::default)
 
-(defmulti visit
-  "Walk a spec definition. Takes two arguments, the spec and the accept
-  function, and returns the result of calling the accept function.
+(defn visit
+  "Walk a spec definition. Takes 2-3 arguments, the spec and the accept
+  function, and optionally a options map, and returns the result of
+  calling the accept function.
 
-  The accept function is called with three arguments: the dispatch term for the
-  spec (see below), the spec itself and the vector with the results of
-  recursively walking the children of the spec.
+  The accept function is called with 4 arguments: the dispatch term for the
+  spec (see below), the spec itself, vector with the results of
+  recursively walking the children of the spec and the options map.
```

Comment:
Would be a good idea to explain what the options map is for. I read the changelog, so I know that this is for, but if I hadn't, based on this docstring I'd think that `visit` itself has some options that aren't explained anywhere.